src/librustc_codegen_ssa/base.rs

```diff
@@ -909,6 +909,39 @@ fn is_codegened_item(tcx: TyCtxt, id: DefId) -> bool {
 }
 
 pub fn provide_both(providers: &mut Providers) {
+    providers.backend_optimization_level = |tcx, cratenum| {
+        let for_speed = match tcx.sess.opts.optimize {
+            // If globally no optimisation is done, #[optimize] has no effect.
+            //
+            // This is done because if we ended up "upgrading" to `-O2` here, we’d populate the
+            // pass manager and it is likely that some module-wide passes (such as inliner or
+            // cross-function constant propagation) would ignore the `optnone` annotation we put
+            // on the functions, thus necessarily involving these functions into optimisations.
+            config::OptLevel::No => return config::OptLevel::No,
+            // If globally optimise-speed is already specified, just use that level.
+            config::OptLevel::Less => return config::OptLevel::Less,
+            config::OptLevel::Default => return config::OptLevel::Default,
+            config::OptLevel::Aggressive => return config::OptLevel::Aggressive,
+            // If globally optimize-for-size has been requested, use -O2 instead (if optimize(size)
+            // are present).
+            config::OptLevel::Size => config::OptLevel::Default,
+            config::OptLevel::SizeMin => config::OptLevel::Default,
+        };
```

Comment:
If I got this right, when we're optimizing for Size or SizeMin and there is a `optimize(speed)` attribute, then this code switches us to create a PM for `OptLevel = 2` and `SizeLevel = 0`, instead of `OptLevel = 2` and `SizeLevel = 1`/`2`. What's the motivation for that?

It shouldn't make much of a difference either way (PM construction depends on `SizeLevel` only in a few minor ways -- ideally those should be switched to check for optsize attributes), just wondering why this particular behavior was picked. Without it, we could save the backend_optimization_level concept.

Reply:
The idea is that we it should likely be the default to have `-Copt-level=s` (for those who care about size) and explicitly annotate `#[optimize(speed)]` where profiling indicates that it is necessary.

In this scenario we want to make sure that we avoid pessimizing functions annotated with `#[optimize(speed)]` to the greatest extent feasible. Hence, if a `#[optimize(size)]` function is found in a module, we instead switch over to `-Copt-level=2` and hope that annotating all the functions with `optsize/minsize` will have mostly the same effect that `-Copt-level=s` had previously with no `#[optimize(size)]` attributes in the crate.

Reply:
(FWIW a possible extension to the RFC was mentioned that we could make it possible to specify both the "speed" level and "size" level, in which case this code would be switched to use the specified "speed" and "size" levels where necessary)